tourcms.js

```diff
@@ -1,799 +1,841 @@
+        let endpoint = TourCMS.TOUR_PICKUP_ROUTES_SHOW + '?id=' + tourId
+        return this.request(endpoint, channel)
+    }
+
+    // Auxiliary method
+
+    generateSignature(path, channelId, verb, outboundTime) {
+        let stringToSign = channelId + '/' + this.marketplaceId + '/' + verb + '/' + outboundTime + path
+        console.log(stringToSign)
+        let signature = rawurlencode(b64_hmac_sha256(this.APIKey, stringToSign) + '=')
+
+        return signature
+    }
+
+    validateParams(params) {
+        if (!params || typeof (params) !== 'string') {
+            return '';
+        }
+
+        if (params && params.substring(0, 1) !== '?') {
+            params = '?' + params;
+        }
+
+        return params;
+    }
+
+    time() {
+        return Math.floor(new Date().getTime() / 1000)
+    }
+
+    parseXml(text) {
+        let xmlDoc = ''
+        if (window.DOMParser) {
+            let parser = new DOMParser()
+            xmlDoc = parser.parseFromString(text, 'text/xml')
+        } else {
+            xmlDoc = new ActiveXObject('Microsoft.XMLDOM')
+            xmlDoc.async = 'false'
+            xmlDoc.loadXML(text)
+        }
+        return xmlDoc
+    }
+
+    xmlToJson(xml) {
+        let attr,
+            child,
+            attrs = xml.attributes,
+            children = xml.childNodes,
+            key = xml.nodeType,
+            obj = {},
+            i = -1
+
+        if (key == 1 && attrs.length) {
+            obj[(key = '@attributes')] = {}
+            while ((attr = attrs.item(++i))) {
+                obj[key][attr.nodeName] = attr.nodeValue
+            }
+            i = -1
+        } else if (key == 3) {
+            obj = xml.nodeValue
+        }
+        while ((child = children.item(++i))) {
+            key = child.nodeName
+            if (obj.hasOwnProperty(key)) {
+                if (obj.toString.call(obj[key]) != '[object Array]') {
+                    obj[key] = [obj[key]]
+                }
+                obj[key].push(this.xmlToJson(child))
+            } else {
+                obj[key] = this.xmlToJson(child)
+            }
+        }
+        return obj
+    }
+
+    setLastResponseHeaders(headers) {
+        let arr = headers.trim().split(/[\r\n]+/)
+        this.lastResponseHeaders = arr
+    }
+
+    getLastResponseHeaders() {
+        return this.lastResponseHeaders
+    }
+
+    // Used for validating webhook signatures
+    validateXMLHash(xml) {
+
+        return this.generateXMLHash(xml) == xml.signed.hash;
+
+    }
+
+    generateXMLHash(xml) {
+
+        let algorithm = xml.signed.algorithm;
+
+        let fields = xml.signed.hash_fields.split(" ");
+
+        let values = [];
+
+        for (let field of fields) {
+
+            let xpathResult = xml.xpath(field);
+
+            for (let result of xpathResult) {
+                values.push(result[0].toString())
+            }
+        }
+
+        let stringToHash = values.join("|");
```

Comment:
ideally would be a default value and a param to override it, it would allow us more flexibility

Reply:
this function is intended to be used with xml object with signature algorithm, and fields